cpp/src/arrow/python/python_to_arrow.cc

```diff
@@ -585,6 +619,37 @@ class PyPrimitiveConverter<T, enable_if_string_like<T>>
     return Status::OK();
   }
 
+  Result<int64_t> ExtendAsMuchAsPossible(PyObject* values, int64_t size,
```

Comment:
Can you try to reconcile the `PyPrimitiveConverter` for binary and string types? They really look very similar.